src/libtls/tls_server.c

```diff
@@ -930,15 +929,47 @@ METHOD(tls_handshake_t, process, status_t,
 	return NEED_MORE;
 }
 
+/**
+ * Write public key into key share extension
+ */
+bool tls_write_key_share(bio_writer_t **key_share, tls_named_group_t group,
+						 diffie_hellman_t *dh)
+{
+	bio_writer_t *writer;
+	chunk_t pub;
+
+	if (!dh || !dh->get_my_public_value(dh, &pub))
+	{
+		return FALSE;
+	}
+	writer = bio_writer_create(pub.len + 7);
+	writer->write_uint16(writer, group);
+	if (group == TLS_CURVE25519 ||
+		group == TLS_CURVE448)
+	{
+		writer->write_data16(writer, pub);
+	}
+	else
+	{	/* classic format (see RFC 8446, section 4.2.8.2) */
+		writer->write_uint16(writer, pub.len + 1);
+		writer->write_uint8(writer, TLS_ANSI_UNCOMPRESSED);
+		writer->write_data(writer, pub);
+	}
+	free(pub.ptr);
+	*key_share = writer;
```

Comment:
You could also just write `*key_share = writer = bio_writer_create(pub.len + 7);`.